### DISCUSSION TITLE
Is nodenext supported?

### DISCUSSION THREAD

**Discussion by hallsbyra** [vitest-dev/vitest#2303] [Q&A] [answered]
When trying to use vitest from a project that uses "module": "NodeNext" I get the following compilation errors.

> node_modules/vite/dist/node/index.d.ts:7:34 - error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
>
>7 import { ConnectedPayload } from "../../types/hmrPayload";

Is it possible to use vitest in a nodenext project?

**Comment by hallsbyra** [comment, marked as answer]
It's very easy to reproduce. But I'm not overly comfortable in the jungle of JS module handling, so I don't know if this is a bug or if I'm doing something wrong.

`tsconfig.json`
```json
{
    "compilerOptions": {
        "module": "NodeNext"
    }
}
```

`package.json` (relevant part)
```json
{
  "type": "module",
}
```

`main.spec.ts`
```ts
import { describe, it } from 'vitest'
````

Executing `npx tsc` results in a bunch of TS2834 errors. The problem seems to be that `node_modules/vite/dist/node/index.d.ts` imports files without extensions, which is prohibited in `NodeNext`.

This seems related: https://github.com/vitest-dev/vitest/issues/325 However, that change (remove `"type": "module"` from `package.json`) appears to have been since reverted.

**Reply by sheremet-va** [reply]
Vitest bundles correct `d.ts` files. Your errors are coming from Vite, so I suggest you open an issue there

**Reply by hallsbyra** [reply]
Ok, thanks.

**Reply by sheremet-va** [reply]
You can also try adding `skipLibCheck`

**Reply by hallsbyra** [reply]
Yup, that makes the error msg go away! Now I run into next problem:

```
TypeError: morgan.format is not a function
 ❯ node_modules/morgan/index.js:152:8
    150|  */
    151|
    152| morgan.format('combined', ':remote-addr - :remote-user [:date[clf]] ":method :url HTTP/:http-version" :status :res[content-length] ":referrer" ":user-agent"')
       |        ^
    153|
    154| /**
```

I wonder if it's related to this: https://github.com/rollup/plugins/issues/606, since I noticed @vite uses rollup. Don't know. I guess I'll stay with jest until vitest stabilizes. Thanks anyway!

**Reply by sheremet-va** [reply]
Vite doesn't use rollup, when running as a server. You have this issue probably because `morgan` has a wrong bundle, it's pretty common. You can add broken library in `deps.inline`.

**Reply by sheremet-va** [reply]
Ah, I see it's not bundled at all, but relies on obscure feature, when assigning `module.exports` binds it to the function by reference, so then it does `module.exports.format` it also adds it to function, huh. You can create a bug issue in Vitest for this case, very interesting.

**Reply by hallsbyra** [reply]
Yeah I think that was the same problem they had in rollup. I'll file a bug.

**Reply by hallsbyra** [reply]
I can't seem to reproduce it in a minimal repo. Just importing morgan is not enough obviously. It only happens in my main repo.
